tests/integration/upgrade_test.go

```diff
@@ -0,0 +1,132 @@
+)
+
+// ************************************************
+// *** Major scenarios tested by the UpgradeSuite ***
+// Setup
+// - Initially create a cluster from the previous minor release
+// - Upgrade to the current build of Rook to verify functionality after upgrade
+// - Test basic usage of block, object, and file after upgrade
+// Monitors
+// - One mon in the cluster
+// ************************************************
+func TestUpgradeSuite(t *testing.T) {
+	s := new(UpgradeSuite)
+	defer func(s *UpgradeSuite) {
+		HandlePanics(recover(), s.op, s.T)
+	}(s)
+	suite.Run(t, s)
+}
+
+type UpgradeSuite struct {
+	suite.Suite
+	helper    *clients.TestClient
+	op        *TestCluster
+	k8sh      *utils.K8sHelper
+	namespace string
+}
+
+func (s *UpgradeSuite) SetupSuite() {
+	s.namespace = "upgrade-ns"
+	useDevices := true
+
+	s.op, s.k8sh = StartTestCluster(s.T, s.namespace, "bluestore", false, useDevices, 1, installer.Version0_8)
+	s.helper = clients.CreateTestClient(s.k8sh, s.op.installer.Manifests)
+}
+
+func (s *UpgradeSuite) TearDownSuite() {
+	s.op.Teardown()
+}
+
+func (s *UpgradeSuite) TestUpgradeToMaster() {
+	systemNamespace := installer.SystemNamespace(s.namespace)
+
+	// Create block, object, and file storage on 0.8 before the upgrade
+	poolName := "upgradepool"
+	storageClassName := "block-upgrade"
+	blockName := "block-claim-upgrade"
+	podName := "test-pod-upgrade"
+	logger.Infof("Initializing block before the upgrade")
+	setupBlockLite(s.helper, s.k8sh, s.Suite, s.namespace, poolName, storageClassName, blockName, podName)
+	createPodWithBlock(s.helper, s.k8sh, s.Suite, s.namespace, blockName, podName)
+	defer blockTestDataCleanUp(s.helper, s.k8sh, s.namespace, poolName, storageClassName, blockName, podName)
+
+	logger.Infof("Initializing file before the upgrade")
+	filesystemName := "upgrade-test-fs"
+	createFilesystem(s.helper, s.k8sh, s.Suite, s.namespace, filesystemName)
+	createFilesystemConsumerPod(s.helper, s.k8sh, s.Suite, s.namespace, filesystemName)
+	defer cleanupFilesystemConsumer(s.helper, s.k8sh, s.Suite, s.namespace, filesystemName)
+
+	logger.Infof("Initializing object before the upgrade")
+	objectStoreName := "upgraded-object"
+	runObjectE2ETestLite(s.helper, s.k8sh, s.Suite, s.namespace, objectStoreName, 1)
+
+	// verify that we're actually running 0.8 before the upgrade
+	operatorContainer := "rook-ceph-operator"
+	version, err := k8sutil.GetDeploymentVersion(s.k8sh.Clientset, systemNamespace, operatorContainer, operatorContainer)
+	assert.Nil(s.T(), err)
+	assert.Equal(s.T(), "v0.8.1", version)
+
+	s.simpleWriteAndRead(podName, "pre-upgrade-file")
+
+	// Upgrade to master
+	_, err = s.k8sh.Kubectl("-n", systemNamespace, "set", "image", "deploy/"+operatorContainer, operatorContainer+"=rook/ceph:"+installer.VersionMaster)
+	require.Nil(s.T(), err)
+
+	// verify that the operator spec is updated
+	version, err = k8sutil.GetDeploymentVersion(s.k8sh.Clientset, systemNamespace, operatorContainer, operatorContainer)
```

Comment:
possible race here between the version being updated above and the single check (with no retry) we do here?

Reply:
The spec should be updated immediately so I don't believe there is a race here. It's possible the new pod isn't running yet, but we're just checking the deployment spec rather than the running pod.